File: sol/solution/0001-0100/0029. Divide Two Integers/Solution.rb
# @param {Integer} dividend
# @param {Integer} divisor
# @return {Integer}
def divide(dividend, divisor)
    int_max = (2**31) - 1
    int_min = -(2**31)
    return int_max if dividend == int_min && divisor == -1
    return int_min if dividend == int_min && divisor == 1
    a, b = dividend.abs, divisor.abs
    res = 0
    while a >= b
        temp, multiple = b, 1
        while (temp << 1) <= a
            temp <<= 1
            multiple <<= 1
        end
        a -= temp
        res += multiple
    end
    (dividend > 0) ^ (divisor > 0) ? -res : res
end

File: sol/solution/0001-0100/0029. Divide Two Integers/Solution.erl
-spec divide(integer(), integer()) -> integer().
divide(Dividend, Divisor) ->
    IntMax = (1 bsl 31) - 1,
    IntMin = -(1 bsl 31),
    case {Dividend, Divisor} of
        {IntMin, -1} -> IntMax;
        {IntMin, 1} -> IntMin;
        _ ->
            A = abs(Dividend),
            B = abs(Divisor),
            Res = divide_loop(A, B, 0),
            if (Dividend > 0) xor (Divisor > 0) -> -Res;
               true -> Res
            end
    end.
divide_loop(A, B, Res) when A < B -> Res;
divide_loop(A, B, Res) ->
    {Temp, Multiple} = double_loop(B, A, B, 1),
    divide_loop(A - Temp, B, Res + Multiple).
double_loop(Temp, A, TempAcc, Multiple) ->
    case (TempAcc bsl 1) =< A of
        true -> double_loop(Temp, A, TempAcc bsl 1, Multiple bsl 1);
        false -> {TempAcc, Multiple}
    end.